### ISSUE TITLE
Please expose current cursor position and current character attributes

### ISSUE THREAD

**Issue by diggernet** [lihop/godot-xterm#72]
So first I need the ability to read the current cursor position from the terminal.

A couple of ideas for how it could be done:
- Expose a function from terminal.cpp that returns the cursor position.
- Implement CSI 6 n (DSR) to output CSI r ; c R (CPR).  Perhaps this could be done by having write() return a string (or PoolByteArray).  So if the input contains that sequence (or any other implemented reporting sequence), write() returns the response to its caller.

Both of those would be useful in different ways, but I think either one could do the job.

The other request is a way to read the current character attributes (SGR).  That is, the fg/bg/bold/etc values that are in effect and will be used on the next write().  

As above, this could be done by:
- One or more functions to read the data.
- Implement DCS $ m Pt ST (DECRQSS) to output a CSI SGR string.  Again, this might mean write() returning that data.

Of course, the implementation ideas above are just my uninformed thoughts.  There's a lot I really don't understand about how this is implemented, so you might easily have a better approach.

**Comment by lihop** [comment]
Hi @diggernet, thanks for the suggestions! I'm currently focused on making the library run smoothly with Godot 4, but will consider them.

> So first I need the ability to read the current cursor position from the terminal.

Data from the terminal comes via the `data_sent` signal, so you can get the position via that. For example:
![2024-02-04-185448_626x387_scrot](https://github.com/lihop/godot-xterm/assets/3696783/b80fc802-6ee9-453d-a3a2-6a1ab8d862a2)
Although having to use `yield` could be a bit awkward, is it suitable for you? I haven't used `await` much in Godot 4, but it might be more user friendly.

> Expose a function from terminal.cpp that returns the cursor position.

libtsm has [`get_cursor_x()` and `get_cursor_y()`](https://github.com/lihop/libtsm/blob/d7b39db8180c6df4b0c9f59a0ce7fbcdb11de60e/src/tsm/libtsm.h#L216-L217) that could quite easily be exposed. Maybe as `get_cursor_pos() -> Vector2`?


> write() returns the response to its caller

While this would avoid the awkwardness of using signals, I'm not sure how feasible it would be for `write()` to return data. The underlying libtsm function (`tsm_vte_input()`) returns `void`, instead calling a write callback when it has a response which is where `data_sent` is emitted from. I'm reluctant to make exceptions to this depending on what sequences the input contains.

> Implement DCS $ q m ST (DECRQSS) to output a CSI SGR string.

`DECRQSS` is not implemented in the original libtsm library or the fork that GodotXterm is using, but we could implement it here: https://github.com/lihop/libtsm/blob/d7b39db8180c6df4b0c9f59a0ce7fbcdb11de60e/src/tsm/tsm-vte.c#L1772-L1779.

Taking a quick look at [libtsm.h](https://github.com/lihop/libtsm/blob/master/src/tsm/libtsm.h), I don't see any methods that would easily expose the attributes, but would certainly prioritize adding support for the `DECRQSS` sequence first.

**Comment by diggernet** [comment]
Yes please on Godot 4.  :)  Godot-xterm is actually the only reason I am currently using 3.5.  I suppose I should mention why your project is so preferable to competitors:  Clear separation of Terminal and PTY.  I mean, the editor plugin is awesome, but I don't currently need or want a command shell in a project.  There all I want is a UI control that is a screen emulator I can send data to.

Anyway, back on topic...
I am happy to learn that there is already a mechanism to return data, and that DSR is already impemented.  However, I see a couple problems with the yield approach.  First, isn't there a race condition where the terminal could in theory trigger data_sent before the yield call sets up a listener?  And second, there is a resource leak if you are writing a bunch of strings that may or may not contain a reporting sequence.  Any string without a reporting sequence leaves a hanging yield handle.  The fun bit is that if & when you do finally trigger data_sent, every one of those hanging yields fires.  But both of those are easily solved by using a signal handler instead of yield.

It would be helpful to also have get_cursor_pos(), but I think I can make do for now until #73 is resolved.

Definitely agree there's no pressing need to modify the behavior of write().  Using data_sent should be good enough.

Looking at tsm-vte.c, it appears the data of interest is in cattr.  But as you say, how to expose it is an open question.

**Comment by diggernet** [comment]
I walked through tsm-vte.c for DECRQSS to see how much of the flow is present.  The gaps seem surprisingly small.
The SGR variant of DECRQSS is `DCS $ q m ST`.  Since DCS can be 'Esc P' or 0x90 and ST can be 'Esc \' or 0x9c, the byte sequence can be:
`(90 or 1b 50) 24 71 6d (9c or 1b 5c)`

The flow from that is approximately:

```
state = STATE_GROUND

0x90
	parse_data()
	switch(raw) case 0x90
	do_trans(STATE_DCS_ENTRY, ACTION_NONE)
	entry action ACTION_CLEAR
	do_clear()
	csi_argc = 0
	csi_argv[] cleared
	csi_flags = 0
	***NOTE 1***
	state = STATE_DCS_ENTRY
-OR-
0x1b
	parse_data()
	switch(raw) case 0x1b
	do_trans(STATE_ESC, ACTION_NONE)
	entry action ACTION_CLEAR
	do_clear()
	csi_argc = 0
	csi_argv[] cleared
	csi_flags = 0
	***NOTE 1***
	state = STATE_ESC
0x50
	parse_data()
	switch(state) case STATE_ESC
	switch(raw) case 0x50
	do_trans(STATE_DCS_ENTRY, ACTION_NONE)
	entry action ACTION_CLEAR
	do_clear()
	csi_argc = 0
	csi_argv[] cleared
	csi_flags = 0
	***NOTE 1***
	state = STATE_DCS_ENTRY

0x24
	parse_data()
	switch(state) case STATE_DCS_ENTRY
	switch(raw) case 0x24
	do_trans(STATE_DCS_INT, ACTION_COLLECT)
	ACTION_COLLECT
	do_collect()
	csi_flags |= CSI_CASH
	***NOTE 2***
	state = STATE_DCS_INT

0x71
	parse_data()
	switch(state) case STATE_DCS_INT
	switch(raw) case 0x71
	do_trans(STATE_DCS_PASS, ACTION_NONE)
	entry action ACTION_DCS_START
	***NOTE 3***
	state = STATE_DCS_PASS

0x6d
	parse_data()
	switch(state) case STATE_DCS_PASS
	switch(raw) case 0x6d
	do_trans(STATE_NONE, ACTION_DCS_COLLECT)
	ACTION_DCS_COLLECT
	***NOTE 4***

0x9c
	parse_data()
	switch(raw) case 0x9c
	do_trans(STATE_GROUND, ACTION_EXECUTE)
	exit action ACTION_DCS_END
	***NOTE 5***
	ACTION_EXECUTE
	do_execute() // "nothing to do here"
	state = STATE_GROUND
-OR-
0x1b
	parse_data()
	switch(raw) case 0x1b
	do_trans(STATE_ESC, ACTION_NONE)
	exit action ACTION_DCS_END
	***NOTE 5***
	entry action ACTION_CLEAR
	do_clear()
	csi_argc = 0
	csi_argv[] cleared
	csi_flags = 0
	***NOTE 1***
	state = STATE_ESC
0x5c
	parse_data()
	switch(state) case STATE_ESC
	switch(raw) case 0x5c
	do_trans(STATE_GROUND, ACTION_ESC_DISPATCH)
	ACTION_ESC_DISPATCH
	do_esc() // "nothing to do here"
	state = STATE_GROUND
```

NOTE 1: There needs to be dsc_len and dsc_arg, similar to osc_len and osc_arg, with their values reset in do_clear().  Or perhaps the OSC vars can simply be reused here.

NOTE 2: This is a DCS command, not CSI, but I don't see any reason not to reuse the csi_flags as the code is already set up to do.

NOTE 3: ACTION_DCS_START needs to trigger a function that will store the current char (in this case 'q') somewhere.  That will be used later to distinguish between DCS commands (u, q, t, Q, p, etc).

NOTE 4: ACTION_DCS_COLLECT needs to trigger a do_dcs_collect() similar to do_osc_collect().  Or perhaps the OSC function can simply be reused here.

NOTE 5: ACTION_DCS_END needs to trigger a do_dcs() function.  For DECRQSS, do_dcs() would check that csi_flags == CSI_CASH and the char stored in NOTE 3 is 'q', in which case it would call a dcs_decrqss() function.  That function would check the string stored in NOTE 4 to determine the request.  Finally, seeing the 'm' it would build and return an appropriate SGR string representing the data in cattrs, similar to the existing csi_dsr() function.

I'm working on a PR for the above.  See what you think of it.  But I'm not set up to do any testing, so please don't just trust my work.  :)

**Comment by lihop** [comment]
`get_cursor_pos()` is in the Godot4 branch now (https://github.com/lihop/godot-xterm/pull/60/commits/575c3875b619f8368c57959bdc7404fd6a17112e) and behaving as expected, although that branch is still a fair ways from being ready.

**Comment by diggernet** [comment]
Excellent!  Looking forward to the Godot4 release.
